File: samples/java/functional/src/games/FizzBuzz.java
package games;

import java.util.concurrent.Future;
import java.util.concurrent.TimeUnit;

import com.ibm.streamsx.topology.TStream;
import com.ibm.streamsx.topology.Topology;
import com.ibm.streamsx.topology.context.StreamsContextFactory;
import com.ibm.streamsx.topology.function.Function;
import com.ibm.streamsx.topology.streams.BeaconStreams;

/**
 * Play
 * <a href="https://en.wikipedia.org/wiki/Fizz_buzz">Fizz Buzz</a>.
 *
 */
public class FizzBuzz {

    /**
     * Entry point for a streaming Fizz Buzz!
     */
    public static void main(String[] args) throws Exception {
        Topology topology = new Topology();
        
        // Declare an infinite stream of Long values
        TStream<Long> counting = BeaconStreams.longBeacon(topology);
        
        // Throttle the rate to allow the output to be seen easier
        counting = counting.throttle(100, TimeUnit.MILLISECONDS);
        
        // Print the tuples to standard output
        playFizzBuzz(counting).print();
        
        // At this point the streaming topology (streaming) is
        // declared, but no data is flowing. The topology
        // must be submitted to a StreamsContext to be executed.

        // Since this is an streaming graph with an endless
        // data source it will run for ever
        Future<?> runningTopology = StreamsContextFactory.getEmbedded().submit(topology);
        
        // Run for one minute before canceling.
        Thread.sleep(TimeUnit.MINUTES.toMillis(1));
        
        runningTopology.cancel(true);       
    }
    
    /**
     * Return a stream that plays Fizz Buzz based
     * upon the values in the input stream.
     */
    public static TStream<String> playFizzBuzz(TStream<Long> counting) {
        
        /*
         * Transform an input stream of longs TStream<Long> to a
         * stream of strings TStream<String> that follow
         * the Fizz Buzz rules based upon each value in the
         * input stream.
         */
        TStream<String> shouts = counting.transform(new Function<Long,String>() {
            private static final long serialVersionUID = 1L;

            @Override
            public String apply(Long v) {
                // Skip 0, humans count from 1!
                if (v == 0)
                    return null;
                
                StringBuilder sb = new StringBuilder();
                if (v % 3 == 0)
                    sb.append("Fizz");
                if (v % 5 == 0)
                    sb.append("Buzz");
                
                if (sb.length() == 0)
                    sb.append(Long.toString(v));
                else
                    sb.append("!");
                
                return sb.toString();
            }});
        
        return shouts;
    }

}

File: samples/python/topology/games/fizz_buzz.py
# Licensed Materials - Property of IBM
# Copyright IBM Corp. 2016,2019
from __future__ import print_function
from streamsx.topology.topology import Topology
import streamsx.topology.context
import fizz_buzz_functions


def main():
    """
    Plays Fizz Buzz (https://en.wikipedia.org/wiki/Fizz_buzz)
    
    Example:
        python3 fizz_buzz.py
    Output:
        1
        2
        Fizz!
        4
        Buzz!
        Fizz!
        7
        8
        Fizz!
        Buzz!
        11
        Fizz!
        13
        14
        FizzBuzz!
        ...

    """
    topo = Topology("fizz_buzz")
    
    # Declare a stream of int values
    counting = topo.source(fizz_buzz_functions.int_tuples)
    
    # Print the tuples to standard output
    play_fizz_buzz(counting).print()
    
    # At this point the streaming topology (streaming) is
    # declared, but no data is flowing. The topology
    # must be submitted to a context to be executed.
    
    # execute the topology by submitting to a standalone context
    streamsx.topology.context.submit("STANDALONE", topo)

def play_fizz_buzz(counting):
    """
    Return a stream that plays Fizz Buzz based
    upon the values in the input stream.
    
    Transform an input stream of integers to a
    stream of strings that follows
    the Fizz Buzz rules based upon each value in the
    input stream.
         
    Args:
        counting: input stream
    Returns:
        transformed output stream
    """
    shouts = counting.map(fizz_buzz_functions.fizz_buzz)
    return shouts

if __name__ == '__main__':
    main()
    
